stablehlo/reference/Ops.cpp

```diff
@@ -133,6 +134,345 @@ Tensor evalConvertOp(const Tensor &operand, TensorType resultType) {
   return result;
 }
 
+Tensor evalConvolutionOp(
+    const Tensor &lhs, const Tensor &rhs, ArrayRef<int64_t> windowStrides,
+    SmallVector<SmallVector<int64_t>> padding, ArrayRef<int64_t> lhsDilation,
+    ArrayRef<int64_t> rhsDilation, ArrayRef<bool> windowReversal,
+    Axis inputBatchDimension, Axis inputFeatureDimension,
+    Axes inputSpatialDimensions, Axis kernelInputFeatureDimension,
+    Axis kernelOutputFeatureDimension, Axes kernelSpatialDimensions,
+    Axis outputBatchDimension, Axis outputFeatureDimension,
+    Axes outputSpatialDimensions, int64_t featureGroupCount,
+    int64_t batchGroupCount, TensorType resultType) {
+  Tensor result(resultType);
+
+  if (resultType.getNumElements() == 0) return result;
+
+  Builder builder = mlir::Builder(resultType.getContext());
+  Type i64Type = builder.getI64Type();
+  Type i1Type = builder.getI1Type();
+
+  auto split = [&](Tensor input, int64_t groupSize, Axis splitDimension) {
+    auto getScalarTensor = [&](auto value) {
+      return makeTensor(
+          DenseElementsAttr::get(RankedTensorType::get({}, i64Type), value));
+    };
+    auto splitInputShape = SmallVector<int64_t>(input.getShape());
+    splitInputShape[splitDimension] /= groupSize;
+
+    auto splitInputType =
+        RankedTensorType::get(splitInputShape, input.getElementType());
+    SmallVector<Tensor> splitResults;
+    for (auto idx = 0; idx < groupSize; ++idx) {
+      SmallVector<Tensor> inputStartIndices(input.getRank(),
+                                            getScalarTensor(0L));
+      inputStartIndices[splitDimension] =
+          getScalarTensor(idx * splitInputShape[splitDimension]);
+      Sizes strides(input.getRank(), 1);
+      auto resultTensor =
+          evalDynamicSliceOp(input, inputStartIndices, strides, splitInputType);
+      splitResults.push_back(resultTensor);
+    }
+    return splitResults;
+  };
+  auto flattenPadding = [&](SmallVector<SmallVector<int64_t>> padding) {
+    SmallVector<int64_t> vector;
+    for (auto pair : padding) vector.append(pair.begin(), pair.end());
+    return vector;
+  };
+  auto getDenseIntElementsAttr = [](SmallVector<int64_t> shape,
+                                    Type elementType, auto value) {
+    return DenseIntElementsAttr::get(RankedTensorType::get(shape, elementType),
+                                     value);
+  };
+
+  if (featureGroupCount > 1) {
+    auto flatPadding = flattenPadding(padding);
+    auto lhses = split(lhs, featureGroupCount, inputFeatureDimension);
+    auto rhses = split(rhs, featureGroupCount, kernelOutputFeatureDimension);
+    SmallVector<int64_t> paddingShape{static_cast<int64_t>(padding.size()), 2};
+    SmallVector<Tensor> results;
+    for (auto [left, right] : llvm::zip(lhses, rhses)) {
+      SmallVector<ShapedTypeComponents> inferredConvolutionType;
+      auto convolutionStatus = hlo::inferConvolutionOp(
```

Comment:
With the use of `inferConvolutionOp` and it's interface https://github.com/openxla/stablehlo/blob/63b6f5ab7ab9059a9134a0ee1a49e731ab376c5f/stablehlo/dialect/TypeInference.cpp#L1733,  I am thinking about using the same interface for `evalConvolutionOp`

Benefit:
1. We can get rid of all these wrap/unwrap code for various attributes.
2. The callsite of `evalConvolutionOp` in `eval`, where we have the loop over the ops, would be simplified a loop.

Reply:
What do you think?

Reply:
I think your overall idea makes sense. But the counterargument would be that it comes at the cost of adding boiler plate code to create default parameters under `eval`, and it would be an exception from a related issue #1031  of unwrapping MLIR based classes out of `inferFooOps`. Let's hear from @burmako before I continue with this change.

Reply:
>> it would be an exception from a related issue https://github.com/openxla/stablehlo/issues/1031 of unwrapping MLIR based classes out of inferFooOps

All we need here is a common interface for inferOp* and evalOp*, so as to remove the conversion code. I am perfectly fine with modifying the inferOp* interfaces as we did earlier as well. 

Sure, let's hear from Eugene first.